### ISSUE TITLE
Re-using Builder instance between builds lead to builds with stale file content

### ISSUE THREAD

**Issue by flyingmutant** [systemjs/builder#336]
If
```js
var builder = new Builder({
    paths: {
      'angular2/*': 'node_modules/angular2/*.js',
      'rx': require.resolve('rx/dist/rx.js'),
    },
    packages: {
      'node_modules/angular2': {
        defaultExtension: 'js',
        format: 'cjs',
      },
      'app': {
        defaultExtension: 'js',
        format: 'register',
      }
    },
  });
```

is only used once (at the top of gulpfile), and

```js
  return builder.buildStatic(PATH.tmp + 'bootstrap.js', PATH.tmp + '_bundle.js', {
    sourceMaps: options.minify,
    runtime: false,
  });
```

is used for building (inside gulp-watch), then any gulp-watch induced build after the first one will use the stale (same as for the first build) file content of files under `app/`.

Moving `new Builder` right before `buildStatic` fixes the issue.

systemjs-builder 0.14.7

**Comment by guybedford** [comment]
Yes, the builder maintains an internal cache state. To clear the builder state use `builder.reset()`.

**Comment by flyingmutant** [comment]
Hm, this behavior is a bit unexpected (and currently undocumented, I believe). What is the rationale for default caching model which will cause broken builds? IMHO it will be better to make this an opt-in.

Is there a way for a more fine-grained (file-level) manual cache invalidation?

**Comment by carlosagsmendes** [comment]
@flyingmutant are you bundling from transpiled code or are your transpiling during the bundling process?

**Comment by guybedford** [comment]
This was added today which may be of interest here - https://github.com/systemjs/builder/pull/341.

**Comment by flyingmutant** [comment]
@carlosagsmendes I am bundling from transpiled code.

**Comment by carlosagsmendes** [comment]
Thank you @flyingmutant  and @guybedford. I have a problem when bundling multiple bundles using the same builder instance. 
If I transpile the code beforehand it works fine. If I bundle from the src folder there seems to be an issue with the babelOptions. The following two lines get invalid values throughout multiple bundles. https://github.com/systemjs/builder/blob/master/compilers/esm.js#L129-L130 (i.e. filenameRelative gets a file name which is not even related to the current bundle). 

@guybedford in my case this is a bug since I use the same instance to do several bundles using gulp but outside a gulp watch task - there are no changes to my code between calls.

**Comment by carlosagsmendes** [comment]
Apparently filenameRelative and  sourceFileName are only being updated when using the builder https://github.com/systemjs/builder/blob/master/compilers/esm.js#L129-L130

When using babelTranspile from system.src.js only the options.filename is being updated 
https://github.com/systemjs/systemjs/blob/1bba480be922f55c788f06a9d6091f7575a08b1f/dist/system.src.js#L1155

This is causing some issues since apparently they are sharing babelOptions on the second case we are getting some wrong data that has been set by the builder.

**Comment by guybedford** [comment]
@carlosagsmendes thanks I've added https://github.com/systemjs/builder/commit/cb038c2a6f3aac9ef0c63fee144ede31595d1fbb let me know if that covers this for you.

The invalidation API has been merged now, so marking this as a documentation issue.

**Comment by carlosagsmendes** [comment]
Thks. I replied there

**Comment by onlywei** [comment]
@guybedford I have just upgraded to try the new `invalidate()` API, but it does not seem to do what I want. I have the following code set up:

```
  let watcher;

  function watch(changed) {
    console.log(builder.invalidate('path/to/my/file.js'));
    builder.trace(mainDotJs).then((tree) => {
      if (watcher) {
        watcher.end();
      }

      console.log(tree['path/to/my/file.js'].metadata.originalSource);
      watcher = gulp.watch('/absolute/path/to/my/file.js', watch);
    });
  }
  watch();
```

Every time I change my file, I `console.log()` the output of `builder.invalidate()`, which results in a non-empty array with a path that looks correct, but the second log message which logs the originalSource of my file in the tree is stale and never changes no matter how many times I change it.

**Comment by guybedford** [comment]
@onlywei `builder.setCache({})` would clear all trace stores, so there wouldn't be any state left containing the trace after that. Are you sure the file contents are being modified?

**Comment by onlywei** [comment]
@guybedford Sorry, my example code was slightly inaccurate. Also I've now updated to the newer version (0.14.9) and I've noticed some differences. 

Here's what the code I meant to show you:

```
  let watcher;

  function watch(changed) {
    console.log(builder.invalidate('path/to/my/file.js'));
    builder.trace('path/to/my/mainFile.js').then((tree) => {
      if (watcher) {
        watcher.end();
      }

      console.log(tree['path/to/my/file.js'].source);
      watcher = gulp.watch('/absolute/path/to/my/file.js', watch);
    });
  }
  watch();
```

The difference between my previous example is that I am calling `builder.trace('mainFile.js')` - where `mainFile.js` imports the file that I am watching for changes.

Previously, in 0.14.8 I was doing the above, and my `console.log()` statement would log the original source of the file correctly, but once I modified the file, it would continue to log the unchanged original source.

Now, in 0.14.9, changing the file causes my `console.log()` statement to log an empty string when I change the file. I just tried `builder.setCache({})` again, and this is also logging empty string when I modify my file.
